clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```diff
@@ -571,7 +571,7 @@ public void close(final long timeoutMs) {
         try {
             maybeAutoCommitOffsetsSync(timeoutMs);
             now = time.milliseconds();
-            if (pendingAsyncCommits.get() > 0 && endTimeMs > now) {
+            if (pendingAsyncCommits.get() > 0 && now - endTimeMs < 0) {
```

Comment:
We've added a `remainingTimeAtLeastZero` function in multiple classes of consumer, maybe we can use that function?

Reply:
I think it is better to use it in line 575 to protect the call to `ensureCoordinatorReady` but not for the if guard due to the strict comparator. If it is correct to write 
```java
now - endTimeMs <= 0
```
then I can replace it with 
```java
remainingTimeAtLeastZero(now, endTimeMs) == 0
```

Reply:
Ack, makes sense.

Reply:
EDIT: thinking about this again, I think we need to consider two cases separately

1) for a parameter passed into a function call like `poll()` higher in the call trace like line 575 below, we need to make sure it is not a negative value, and hence it's better to use `remainingTimeAtLeastZero`.

2) for a condition comparing two values, we need to first of all make sure none of the comparing values may overflow, for example in `KerberosLogin.java` below if `now + minTimeBeforeRelogin` already overflows, then `expiry - (now + minTimeBeforeRelogin) < 0` is not safe either. So I think a better solution would be, for any comparing values we makes sure they are directly from timer than as a result of sum, before going on to the condition check: for example in `NetworkClientUtils.java` above, the comparing `expiryTime` is a result of `startTime + timeoutMs`, instead, we should refactor it as:

```
long startTime = time.milliseconds();

// some calls that may take time

long attemptStartTime = time.milliseconds();

while (!client.isReady(node, attemptStartTime) && attemptStartTime - startTime < timeoutMs)
```

WDYT?

Reply:
For the first point I'm totally agreed with you. 
Regarding the second instead,  the inequation `expiry - (now + minTimeBeforeRelogin) < 0` is still safe also if `now + minTimeBeforeRelogin` already overflowed. The critical bit with a timestamp comparison is to compare always with a constant/safe value. 
Due to the finite arithmetic of computers, the safest way to compare timestamp variables is to perform operations on them only on one side of the inequation. I created a tool that automatically identifies the critical program points and performs these changes. Therefore, I use `0` as safe value.  However, your suggestion of comparing values directly from timers such as `attemptStartTime - startTime < timeoutMs` is even better because it is more natural and readable.  

If you agree, I would go though the code again and apply the changes according to the two cases.

Reply:
Sounds good, thanks.